File: JavaScript/word-count.js
function removePunctuation(str) {
    let punctuation = [".", "\n", "?", ":", ",", "!", "&", "@", "$", "%", "^", "&"];

    punctuation.forEach(x => {
        str = str.replaceAll(x, " ");
    });

    return str;
}

function clean(str) {
    // Remove leading and trailing quotes if present
    str = str.replace(/^['"]|['"]$/g, '');
    return str.toLowerCase().trim();
}

function count(str) {
    str = removePunctuation(str);
    let words = str.split(" ").map(x => clean(x)).filter(x => x != "");

    const counter = {};

    for (const word of words) {
        counter[word] = (counter[word] || 0) + 1;
    }

    return counter;
}

const inputs = [
    "reserved words like constructor and toString ok?",
    " multiple   whitespaces",
    "First: don't laugh. Then: don't cry. You're getting it.",
    "go Go GO Stop stop",
    ",\n,one,\n ,two \n 'three'"
];

inputs.forEach(input => {
    console.log(count(input));
});

File: C#/Word Count/WordCount.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.Linq;

public static class WordCount
{
    public static string RemovePunctuation(string str)
    {
        string[] punctuation = [".", "\n", "?", ":", ",", "!", "&", "@", "$", "%", "^", "&"];

        foreach(var c in punctuation)
        { 
            str = str.Replace(c, " ");
        };

        return str;
    }

    public static string Clean(string str)
    { 
        // Remove leading and trailing quotes if present
        str = Regex.Replace(str, @"^['""]|['""]$", "");

        return str.ToLower().Trim();
    }

    public static IDictionary<string, int> CountWords(string phrase)
    {
        var d = new Dictionary<string, int>();

        var words = RemovePunctuation(phrase).Split(" ")
                    .Select(x => Clean(x))
                    .Where(x => x != "") 
                    .ToArray();

        foreach (var word in words)
        {
            if (d.ContainsKey(word))
            {
                d[word] += 1;
            }
            else
            {
                d[word] = 1;
            }
        }

        return d;
    }
}